python/cudf/cudf/core/mixins/mixin_factory.py

```diff
@@ -0,0 +1,256 @@
+    def wrapper(self, *args2, **kwargs2):
+        return method(self, *args1, *args2, **kwargs1, **kwargs2)
+
+    return wrapper
+
+
+class Operation:
+    """Descriptor used to define operations for delegating mixins.
+
+    This class is designed to be assigned to the attributes (the delegating
+    methods) defined by the OperationMixin. This class will create the method
+    and mimic all the expected attributes for that method to appear as though
+    it was originally designed on the class. The use of the descriptor pattern
+    ensures that the method is only created the first time it is invoked, after
+    which all further calls use the callable generated on the first invocation.
+
+    Parameters
+    ----------
+    name : str
+        The name of the operation.
+    docstring_format_args : str
+        The attribute of the owning class from which to pull format parameters
+        for this operation's docstring.
+    base_operation : str
+        The underlying operation function to be invoked when operation `name`
+        is called on the owning class.
+    """
+
+    def __init__(self, name, docstring_format_args, base_operation):
+        self._name = name
+        self._docstring_format_args = docstring_format_args
+        self._base_operation = base_operation
+
+    def __get__(self, obj, owner=None):
+        retfunc = _partialmethod(self._base_operation, op=self._name)
+
+        # Required attributes that will exist.
+        retfunc.__name__ = self._name
+        retfunc.__qualname__ = ".".join([owner.__name__, self._name])
+        retfunc.__module__ = self._base_operation.__module__
+
+        if self._base_operation.__doc__ is not None:
+            retfunc.__doc__ = self._base_operation.__doc__.format(
+                cls=owner.__name__,
+                op=self._name,
+                **self._docstring_format_args,
+            )
+
+        retfunc.__annotations__ = self._base_operation.__annotations__.copy()
+        retfunc.__annotations__.pop("op", None)
+        retfunc_params = [
+            v
+            for k, v in inspect.signature(
+                self._base_operation
+            ).parameters.items()
+            if k != "op"
+        ]
+        retfunc.__signature__ = inspect.Signature(retfunc_params)
+
+        setattr(owner, self._name, retfunc)
+
+        if obj is None:
+            return getattr(owner, self._name)
+        else:
+            return getattr(obj, self._name)
+
+
+def _should_define_operation(cls, operation, base_operation_name):
+    if operation not in dir(cls):
+        return True
+
+    # If the class doesn't override the base operation we stick to whatever
+    # parent implementation exists.
+    if base_operation_name not in cls.__dict__:
+        return False
+
+    # There are two possibilities here:
+    # 1. A parent class manually overrode the operation. That override takes
+    #    precedence even if the current class overrode the base_operation.
```

Comment:
This is ok, as long as implementation of the parent class respects dynamic dispatching to the correct `_base_operation`.